### ISSUE TITLE
[3.0.1] https://yandex.ru 404 through proxy

### ISSUE THREAD

**Issue by Sobes76rus** [aio-libs/aiohttp#2736]
Hello. 

```
async with aiohttp.ClientSession() as client:
    async with client.get('https://yandex.ru', proxy=proxy) as resp:
        assert resp.status == 200
```

I try to get response from https://yandex.ru through proxy, but response status code is 404
`aiohttp 3.0.1`

When i use `aiohttp 2.3.10` this problem disappear

**Comment by asvetlov** [comment]
I don't recall any proxy related difference between 2.3 and 3.0
The regression is still possible but I need your help for figuring out the problem.
Proxy support is very delicate thing: I cannot reproduce results given by **your** unknown proxy.

What can help is:
1. Grabbing sent/received traffic and calculating the difference and source of rejection 
2. Sharing the proxy access (you can do it privately, no need to show the data to entire internet)

**Comment by Sobes76rus** [comment]
I checked this for different proxy servers, including squid on localhost, but result still the same
I will try to grab traffic in the evening

**Comment by webknjaz** [comment]
@Sobes76rus we still need a traffic dump or a PR with testcase breaking the flow.

**Comment by Sobes76rus** [comment]
I have sniffed requests to the https://yandex.ru/ (each request was made over TLS) and decrypted both of them.

As you can see the only difference in URI in starting line (and in the User-Agent header but I don't think this is the reason).

Below you can see decrypted requests:

aiohttp 3.0.1 decrypted request
```
GET https://yandex.ru HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Host: yandex.ru
User-Agent: Python/3.6 aiohttp/3.0.1
Connection: close
```
response
```
<ClientResponse(https://yandex.ru) [404 Not Found]>
<CIMultiDictProxy('Date': 'Wed, 14 Feb 2018 17:21:38 GMT', 'Content-Type': 'text/html; charset=UTF-8', 'Cache-Control': 'no-cache,no-store,max-age=0,must-revalidate', 'Expires': 'Wed, 14 Feb 2018 17:21:39 GMT', 'Last-Modified': 'Wed, 14 Feb 2018 17:21:39 GMT', 'P3P': 'policyref="/w3c/p3p.xml", CP="NON DSP ADM DEV PSD IVDo OUR IND STP PHY PRE NAV UNI"', 'X-XSS-Protection': '1; mode=block', 'X-Content-Type-Options': 'nosniff', 'Set-Cookie': 'i=BiHiSYaIYSjjJUuT+QM4pEyE2tKggaZVX+dQGSm7BiAu7Y5SRUyxazSti7mbsUI/hrUbzzjVqiLjXlNdbdKohuwUaDA=; Expires=Sat, 12-Feb-2028 17:21:38 GMT; Domain=.yandex.ru; Path=/; Secure; HttpOnly', 'Connection': 'Close', 'Content-Length': '13905')>
```

aiohttp 2.3 decrypted request
```
GET / HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Host: yandex.ru
User-Agent: Python/3.6 aiohttp/2.3.0
Connection: close
```
response 
```
<ClientResponse(https://yandex.ru) [200 Ok]>
<CIMultiDictProxy('Date': 'Wed, 14 Feb 2018 17:22:47 GMT', 'Content-Type': 'text/html; charset=UTF-8', 'Cache-Control': 'no-cache,no-store,max-age=0,must-revalidate', 'Expires': 'Wed, 14 Feb 2018 17:22:47 GMT', 'Last-Modified': 'Wed, 14 Feb 2018 17:22:47 GMT', 'Content-Security-Policy': "connect-src 'self' wss://webasr.yandex.net https://mc.webvisor.com https://mc.webvisor.org wss://push.yandex.ru wss://portal-xiva.yandex.net https://yastatic.net https://home.yastatic.net https://yandex.ru https://*.yandex.ru static.yandex.sx brotli.yastatic.net et.yastatic.net portal-xiva.yandex.net yastatic.net home.yastatic.net yandex.ru *.yandex.ru *.yandex.net yandex.st; default-src 'self' blob: wss://portal-xiva.yandex.net yastatic.net portal-xiva.yandex.net; font-src 'self' https://yastatic.net static.yandex.sx brotli.yastatic.net et.yastatic.net yastatic.net; frame-src 'self' yabrowser: data: https://ok.ru https://www.youtube.com https://player.video.yandex.net https://ya.ru https://yastatic.net https://yandex.ru https://*.yandex.ru wfarm.yandex.net yastatic.net yandex.ru *.yandex.ru awaps.yandex.net *.cdn.yandex.net; img-src 'self' data: https://yastatic.net https://home.yastatic.net https://*.yandex.ru https://*.yandex.net https://*.tns-counter.ru awaps.yandex.net *.yastatic.net gdeua.hit.gemius.pl pa.tns-ua.com mc.yandex.com mc.webvisor.com mc.webvisor.org static.yandex.sx brotli.yastatic.net et.yastatic.net yastatic.net home.yastatic.net yandex.ru *.yandex.ru *.yandex.net *.tns-counter.ru *.gemius.pl yandex.st; media-src 'self' blob: data: *.storage.yandex.net yastatic.net kiks.yandex.ru strm.yandex.ru; object-src 'self' *.yandex.net music.yandex.ru strm.yandex.ru yastatic.net kiks.yandex.ru awaps.yandex.net storage.mds.yandex.net; report-uri https://csp.yandex.net/csp?from=big.ru&showid=1518628967.13664.20932.22703&h=f2&yandexuid=3468659181518628967; script-src 'self' 'unsafe-inline' 'unsafe-eval' blob: https://suburban-widget.rasp.yandex.ru https://suburban-widget.rasp.yandex.net https://music.yandex.ru https://mc.yandex.fr https://mc.webvisor.com https://yandex.fr https://mc.webvisor.org https://yastatic.net https://home.yastatic.net https://mc.yandex.ru https://pass.yandex.ru an.yandex.ru api-maps.yandex.ru static.yandex.sx webasr.yandex.net brotli.yastatic.net et.yastatic.net yastatic.net home.yastatic.net yandex.ru www.yandex.ru mc.yandex.ru suggest.yandex.ru clck.yandex.ru awaps.yandex.net; style-src 'self' 'unsafe-inline' https://yastatic.net https://home.yastatic.net static.yandex.sx brotli.yastatic.net et.yastatic.net yastatic.net home.yastatic.net;", 'P3P': 'policyref="/w3c/p3p.xml", CP="NON DSP ADM DEV PSD IVDo OUR IND STP PHY PRE NAV UNI"', 'Set-Cookie': 'yp=1521220967.ygu.1; Expires=Sat, 12-Feb-2028 17:22:47 GMT; Domain=.yandex.ru; Path=/', 'Set-Cookie': 'mda=0; Expires=Thu, 14-Jun-2018 17:22:47 GMT; Domain=.yandex.ru; Path=/', 'Set-Cookie': 'yandex_gid=10502; Expires=Fri, 16-Mar-2018 17:22:47 GMT; Domain=.yandex.ru; Path=/', 'Set-Cookie': 'yandexuid=3468659181518628967; Expires=Sat, 12-Feb-2028 17:22:47 GMT; Domain=.yandex.ru; Path=/', 'X-Frame-Options': 'DENY', 'Set-Cookie': 'i=A/EU2ql4040jTY2mLWi/VqiPzN7lXBWPud1FgbH0V9viQ2bxauTIuX1yARUR656F8rTtk9g4f6tMiiyD7LgCC1MIwAQ=; Expires=Sat, 12-Feb-2028 17:22:47 GMT; Domain=.yandex.ru; Path=/; Secure; HttpOnly', 'X-Content-Type-Options': 'nosniff', 'Connection': 'Close', 'Content-Length': '137144')>
```

**Comment by asvetlov** [comment]
thanks

**Comment by webknjaz** [comment]
According to RFC, queries to proxy MUST contain absolute-form URI in their Request-Line.
(IIRC)

**Comment by iDeBugger** [comment]
@webknjaz You must be talking about CONNECT query. As far as I know all requests after CONNECT must follow regular rules. Isn't it?

**Comment by webknjaz** [comment]
I need to check RFC, I'm not sure.

**Comment by webknjaz** [comment]
Please also note that last paragraph about servers being required to accept absolute-form even for non-proxy requests for forward compat. Oh and there's smth about domain in request-target taking precedence over Host header for such requests in some other RFCs, but it's not related to current issue.

https://tools.ietf.org/html/rfc7230#section-5.3.2
```
5.3.2.  absolute-form
   When making a request to a proxy, other than a CONNECT or server-wide
   OPTIONS request (as detailed below), a client MUST send the target
   URI in absolute-form as the request-target.

     absolute-form  = absolute-URI

   The proxy is requested to either service that request from a valid
   cache, if possible, or make the same request on the client's behalf
   to either the next inbound proxy server or directly to the origin
   server indicated by the request-target.  Requirements on such
   "forwarding" of messages are defined in Section 5.7.

   An example absolute-form of request-line would be:

     GET http://www.example.org/pub/WWW/TheProject.html HTTP/1.1

   To allow for transition to the absolute-form for all requests in some
   future version of HTTP, a server MUST accept the absolute-form in
   requests, even though HTTP/1.1 clients will only send them in
   requests to proxies.
```

@iDeBugger CONNECT requires authority-form:
https://tools.ietf.org/html/rfc7230#section-5.3.3
```
5.3.3.  authority-form

   The authority-form of request-target is only used for CONNECT
   requests (Section 4.3.6 of [RFC7231]).

     authority-form = authority

   When making a CONNECT request to establish a tunnel through one or
   more proxies, a client MUST send only the target URI's authority
   component (excluding any userinfo and its "@" delimiter) as the
   request-target.  For example,

     CONNECT www.example.com:80 HTTP/1.1
```

**Comment by Sobes76rus** [comment]
After the CONNECT query to proxy and establish connection to desctination server, proxy became a tunnel.
According to [RFC](https://tools.ietf.org/html/rfc7230#section-2.3):
```
A "tunnel" acts as a blind relay between two connections without
   changing the messages.  Once active, a tunnel is not considered a
   party to the HTTP communication, though the tunnel might have been
   initiated by an HTTP request.
```
This can be meant as direct connection

To check this theory I have sniffed browser's requests through proxy to the https://yandex.ru/ (each request was made over TLS) and decrypted them

Google Chrome
```
GET / HTTP/1.1
Host: yandex.ru
Connection: keep-alive
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/64.0.3282.119 Safari/537.36
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
Cookie: ...
```

Firefox
```
GET / HTTP/1.1
Host: yandex.ru
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:59.0) Gecko/20100101 Firefox/59.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: ru-RU,ru;q=0.8,en-US;q=0.5,en;q=0.3
Accept-Encoding: gzip, deflate, br
Cookie: ...
Connection: keep-alive
Upgrade-Insecure-Requests: 1
Cache-Control: max-age=0
```

As you can see, both of them use `origin-form`

**Comment by webknjaz** [comment]
> When making a request to a proxy, other than a CONNECT 

Makes me think that there's an ad-hoc mode of requesting resource from proxy w/o CONNECTing

**Comment by webknjaz** [comment]
Apparently tunneling proxies are used for TLS termination, while simplier ones don't use CONNECT: https://parsiya.net/blog/2016-07-28-thick-client-proxying---part-6-how-https-proxies-work/

Still article refers to attaching `Host` header for backwards compatibility with HTTP/1.0

**Comment by webknjaz** [comment]
That said, if the proxy uses CONNECT to set up a tunnel, it should use origin-form + Host header, otherwise absolute-form is required.
Any objections?

**Comment by webknjaz** [comment]
Basically, CONNECT instructs your proxy to set up a TCP connection with some third-party host and can be used for forwarding of any bytes through such connection.
Ref: https://stackoverflow.com/a/11698002/595220 + https://stackoverflow.com/a/40329026/595220

**Comment by webknjaz** [comment]
Oh, and this implies proxy chaining

**Comment by webknjaz** [comment]
Wiki: https://en.wikipedia.org/wiki/HTTP_tunnel#HTTP_CONNECT_method

**Comment by asvetlov** [comment]
We should write functional tests for tunneled proxies.
Right now tests are just disabled :(

**Comment by asvetlov** [comment]
Should be fixed by aiohttp==3.0.7

**Comment by Sobes76rus** [comment]
Thx =)
